packages/transformers/babel/src/config.js

```diff
@@ -27,18 +28,49 @@ export default async function getBabelConfig(asset: MutableAsset) {
   let result = {};
   mergeConfigs(result, babelrc);
 
-  // Add a generated babel-preset-env config if it is not already specified in the babelrc
-  let hasEnv =
-    babelrc &&
-    hasPlugin(babelrc.config.presets, [
-      'env',
-      '@babel/env',
-      '@babel/preset-env'
-    ]);
+  // Typescript must use the plugin directly (not the typescript preset) and must
+  // come before preset env, otherwise proposals and nonstandard syntax is not
+  // transformed in time.
+  if (path.extname(asset.filePath).match(TYPESCRIPT_EXTNAME_RE)) {
+    let hasTypescript =
+      babelrc &&
+      (hasPlugin(babelrc.config.presets, [
+        '@babel/typescript',
+        '@babel/preset-typescript'
+      ]) ||
+        hasPlugin(babelrc.config.plugins, [
+          '@babel/transform-typescript',
+          '@babel/plugin-transform-typescript'
+        ]));
```

Comment:
I don't really think we should be doing this kind of searching for babel plugins and manipulation of user config anymore. It's not safe, and not possible to do fully correctly especially if we are supporting `babel.config.js`, but also if someone used e.g. a forked version of the typescript preset or something. I think we should just pull this into a separate parcel transformer to just run babel-preset-typescript. If it happens to run again as part of the babel transformer later on, it shouldn't break anything (ts is already gone).

Reply:
I was also thinking of splitting the babel-preset-env and flow steps into their own parcel plugins too, and having the babel transformer only run if there is an explicit `.babelrc` config... not sure yet.

Reply:
> I don't really think we should be doing this kind of searching for babel plugins and manipulation of user config anymore

I agree with this, Jest does it and I've had to debug that behavior in Jest several times and it's been a complete nightmare that takes several days to work out every time.